shtools: new port

MarkWieczorek opened the pull request:
#### Description

shtools is an archive of Fortran 95 software for performing spherical harmonic transforms and related operations. It is mature software for more than a decade and is an integral component of the python project pyshtools. More information can be found here: https://shtools.github.io/SHTOOLS/

###### Type(s)
submission (new Portfile)

###### Tested on
macOS 10.15.6 19G2021
Xcode 12.0 12A7209

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?
<!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

MarkWieczorek commented:
For reference, here is previous submission that required modifying the makefiles to work with `gfortran-mp-10`: https://github.com/macports/macports-ports/pull/8424

chrstphrchvz commented on science/shtools/Portfile:
```diff
@@ -0,0 +1,48 @@
+
+depends_build       port:gcc10 port:fftw-3
+depends_test        port:gcc10 port:fftw-3
+depends_lib         port:fftw-3
```
I personally don't know the answer, but should this port also specify `libgcc10` as a library dependency?

MarkWieczorek replied:
I have never had to specify `libgcc10` on linux or macOS before, and the builds below are passing. So my guess is No. The only reason we need `gcc10` at all is to install the `gfortran` compiler.

ryandesign replied:
> the builds below are passing. So my guess is No.

A successful CI build by no means guarantees that everything in a portfile is correct. But in this case it looks like the only binary file this port installs is the static library /opt/local/lib/libSHTOOLS.a. As such I agree that it doesn't have a library dependency libgcc10. And parts of libgcc10 that it needs will have been included within the static library.

ryandesign replied:
According to the instructions that the build prints, to use the library, one must link with `-lSHTOOLS -lfftw-3`. So the library dependency on fftw-3 is justified. And therefore, fftw-3 does not need to be specified as a build or test dependency; library dependencies are guaranteed to be present at build and test time.

MarkWieczorek commented:
Sorry for the new commit: 

When installing with the option `+openmp`, parallel builds are not supported. (Parallel builds work fine with the default install options.)

MarkWieczorek commented:
All the tests above passed, but if you look at the Travis logs in detail, you'll see that there is one failure for xcode9.4. This failure seems to be unrelated to the new formula (perhaps this is a known issue), so someone might want to try re-running that job to see what happens.

chrstphrchvz commented:
The Travis CI Xcode 9.4 failure can be ignored. It was because a dependency could not be installed, due to a recurring error on Travis CI:
```
Portfile is from the future - check date and time of your system
```
This has been investigated with no working solution yet. Time limits and network reliability issues already limit Travis CI's usefulness for MacPorts; it has been considered to stop using it.

ryandesign commented on science/shtools/Portfile:
```diff
@@ -0,0 +1,49 @@
+
+categories          science math
+platforms           darwin
+license             BSD-3-Clause
```
In MacPorts we call this license "BSD". See https://trac.macports.org/wiki/PortfileRecipes#license. We don't use SPDX identifiers yet. See https://trac.macports.org/ticket/61237.

ryandesign commented on science/shtools/Portfile:
```diff
@@ -0,0 +1,49 @@
+github.tarball_from   archive
+checksums           sha256  176610743ddd76ea1fe7b45db3e8613171630b85c3759c9a4b31cde6b5d2ae3d \
+                    rmd160  58959a91c2889488844f6a7e4745090b0c906057 \
+                    size    35165653
```
```
$ port destr --
--->  Computing dependencies for shtools
--->  Fetching distfiles for shtools
--->  Attempting to fetch SHTOOLS-4.7.1.tar.gz from https://distfiles.macports.org/shtools
--->  Attempting to fetch SHTOOLS-4.7.1.tar.gz from https://github.com/SHTOOLS/SHTOOLS/archive/v4.7.1
--->  Verifying checksums for shtools
Error: Checksum (sha256) mismatch for SHTOOLS-4.7.1.tar.gz
Error: Checksum (rmd160) mismatch for SHTOOLS-4.7.1.tar.gz
Error: Checksum (size) mismatch for SHTOOLS-4.7.1.tar.gz
Error: Failed to checksum shtools: Unable to verify file checksums
```

ryandesign commented on science/shtools/Portfile:
```diff
@@ -0,0 +1,49 @@
+test.args           PREFIX=${prefix} F95="${prefix}/bin/gfortran-mp-10" LAPACK="-framework Accelerate" BLAS=""
+
+destroot.cmd        make
+destroot.target     install
```
I think you can remove the lines beginning build.cmd, test.cmd, destroot.cmd and destroot.target, because the values you've specified are the default values.

MarkWieczorek commented:
Thanks for the comments.

* I changed BSD-3-Clause to just BSD.
* I removed the redundant lines specifying default parameters for build.cmd, test.cmd, destroot.cmd and destroot.target.
* I removed the fftw-3 from the build and test dependencies.
* I updated the hashes, **but I don't understand why these changed**. You can verify in the logs above that the hashes matched when I submitted this PR. A diff of the original file that I saved on my computer and the one I downloaded today shows
```
diff -r SHTOOLS-4.7.1/pyshtools/_version.py SHTOOLS2-4.7.1/pyshtools/_version.py
26c26
<     git_refnames = " (HEAD -> master, tag: v4.7.1)"
---
>     git_refnames = " (HEAD -> master, tag: v4.7.1, develop)"
```
So there is some issue with a tag.`develop` is the main branch that we work on and the python package versioneer creates a static version number from the git tag. However, the only thing that would make sense is if github **recreated** the tar.gz file after I downloaded the first version. This is really bizarre and could break things for other projects besides mine. If you have any insight as to what is going on, please let me know!!!

MarkWieczorek commented:
For info, I have opened an issue at GitHub support about the hash of the release changing. The more I think about this, the more I think that there is a major problem.

chrstphrchvz commented:
You are encountering a behavior affecting projects using python-versioneer which has been known about for some time, so it is unlikely to be something GitHub is directly responsible for. (An old example MacPorts encountered: https://trac.macports.org/ticket/41657#comment:2. Other examples: https://github.com/spack/spack/issues/7937, https://github.com/jborg/attic/issues/335 ) I am not familiar with python-versioneer, but from what I can tell, some workarounds are to avoid using `git_refnames`; or to be diligent in your workflow to make sure that the development branches contain/descend from any commits to be tagged, and that the development branches be pushed to the remote before pushing new tags to the remote.
